pgx/src/shmem.rs

```diff
@@ -0,0 +1,124 @@
+use core::ops::{Deref, DerefMut};
+use crate::pg_sys;
+use std::cell::UnsafeCell;
+
+// RwLock clone which uses a PostgreSQL LWLock to guard data (so it's safe cross process)
+pub struct PgLwLock<T> {
+    inner: UnsafeCell<Option<PgLwLockInner<T>>>,
+}
+
+impl<T> PgLwLock<T> {
+    pub const fn new() -> Self {
+        PgLwLock {
+            inner: UnsafeCell::new(None),
+        }
+    }
+    pub fn share(&self) -> PgLwLockShareGuard<T> {
+        unsafe { (*self.inner.get()).as_ref().unwrap().share() }
+    }
+
+    pub fn exclusive(&self) -> PgLwLockExclusiveGuard<T> {
+        unsafe { (*self.inner.get()).as_mut().unwrap().exclusive() }
+    }
+
+    // To use a lock it has to be attached to an allocated Postgres LWLock, which
+    // we look up by name
+    pub fn attach(&self, lock: String, value: T) -> Result<(), T> {
```

Comment:
why not just make this the `new` constructor instead?

Reply:
Generally (at least my usage so far) the lock gets created (but not attached) as a global, then attached later on.

let me think about this a bit more